addons/web/models/models.py

```diff
@@ -392,28 +421,78 @@ def search_panel_select_range(self, field_name, **kwargs):
             raise UserError(_('Only types %(supported_types)s are supported for category (found type %(field_type)s)') % ({
                             'supported_types': supported_types, 'field_type': field.type}))
 
-        field_names = {}
+        model_domain = AND([
+            kwargs.get('search_domain', []),
+            kwargs.get('category_domain', []),
+            [(field_name, '!=', False)]
+        ])
+
+        if field.type == 'selection':
+            return {
+                'parent_field': False,
+                'values': self._search_panel_selection_range(field_name, model_domain=model_domain,**kwargs),
+            }
+
+        enable_counters = kwargs.get('enable_counters')
+        expand = kwargs.get('expand')
+        domain_image = {}
+        if enable_counters or not expand:
+            domain_image = self._search_panel_domain_image(field_name, model_domain, enable_counters)
+
+        Comodel = self.env[field.comodel_name].with_context(hierarchical_naming=False)
+        field_names = ['display_name']
+        hierarchize = kwargs.get('hierarchize', True)
         parent_name = False
-        if field.type == 'many2one':
-            Comodel = self.env[field.comodel_name]
-            if Comodel._parent_name in Comodel._fields:
-                parent_name = Comodel._parent_name
+        if hierarchize and Comodel._parent_name in Comodel._fields:
+            parent_name = Comodel._parent_name
+            field_names.append(parent_name)
 
-                def process_parent_field(record):
-                    parent_id = record[parent_name]
-                    if parent_id:
-                        record[parent_name] = parent_id[0]
+            def get_parent_id(record):
+                value = record[parent_name]
+                return value and value[0]
+        else:
+            hierarchize = False
 
-                field_names[parent_name] = process_parent_field
+        comodel_domain = kwargs.get('comodel_domain', [])
+        if not (expand or hierarchize or comodel_domain):
+            return {
+                'parent_field': parent_name,
+                'values': list(domain_image.values()),
+            }
 
-        values_range = self._search_panel_range(field_name, comodel_field_names=field_names, **kwargs)
+        if not expand:
+            image_element_ids = list(domain_image.keys())
+            if hierarchize:
+                condition = [('id', 'parent_of', image_element_ids)]
+            else:
+                condition = [('id', 'in', image_element_ids)]
+            comodel_domain = AND([comodel_domain, condition])
+        comodel_records = Comodel.search_read(comodel_domain, field_names)
+
+        if hierarchize:
+            ids = map(lambda rec: rec['id'], comodel_records) if expand else image_element_ids
```

Comment:
Use a list comprehension instead
```suggestion
            ids = [rec['id'] for rec in comodel_records] if expand else image_element_ids
```